morf-core/src/main/java/org/alfasoftware/morf/upgrade/Upgrade.java

```diff
@@ -334,6 +341,35 @@ long getUpgradeAuditRowCount(ResultSetProcessor<Long> processor) {
   }
 
 
+  /**
+   * Generates SQL to be run at the start of the upgrade script which ensures the upgrade can't be run twice
+   */
+  private List<String> getOptimisticLockingInitialisationSql(long upgradeAuditCount) {
+    TableReference upgradeStatusTable = tableRef(UpgradeStatusTableService.UPGRADE_STATUS);
+
+    SelectStatement selectStatement = select().from(upgradeStatusTable)
+        .where(neq(selectUpgradeAuditTableCount().asField(), upgradeAuditCount))
+        .build();
+
+    InsertStatement insertStatement = insert().into(upgradeStatusTable)
+        .from(selectStatement)
+        .build();
+
+    return connectionResources.sqlDialect().convertStatementToSQL(insertStatement);
+  }

```

Comment:
Because all the tests are mocked, we do not actually test your new code on an actual database.
We do not therefore test positive and negative scenarios.

For example, this locking relies on zzzUpgradeStatus having both a primary key and at least one record.
If someone reworks the logic around this table and either removes the primary key or forgets to insert at least one record, your lock will still happily execute, but it will stop locking.

At the minimum, we have to ensure the zzzUpgradeStatus table, created in UpgradeStatusTableServiceImpl.updateTableScript(), is doing both the PK and the insert. But better than that, we should have a test where UpgradeStatusTableService and Upgrade are shown to do the locking, and prove that the lock actually works and allows/stops the upgrade in different scenarios.

At the moment, your locking is intertwined with this Upgrade class, and therefore must be tested within the entire Upgrade logic, which is perhaps not too bad, and we might have such non-mocked test already, which you can use to test this.

Alternatively, extracting this locking into a new class would allow you to do something as simple as TestDatabaseUpgradeLockService, just to make sure that your locking and UpgradeStatusTableService work as expected.